zonemaster/zonemaster-backend
# Correct the handling of delegated and undelegated in historic tests

## Purpose

The feature to separate delegated from undelegated tests (and vice versa) when listing historic tests does not work due to incorrect implementation.

## Context

Fixes #812.

## Changes

* Creates a clear specification in the API documentation of the meaning of "delegated" vs "undelegated".
* Creates a shared method in DB.pm that calculates if the test is delegated or undelegated.
* Updates the implementation in SQLite.pm and MySQL.pm to make it work and follow the specification. Also for tests started with `add_batch_job`.
* Updates the database schema for PostgreSQL to follow the other database engines and to make it possible implement all engines with the same structure.  
* Updates the implementation in PostgreSQL.pm to make it work and follow the specification. Also for tests started with `add_batch_job`.
* Adds instructions and script for updating a PostgreSQL database.

Since this PR changes the schema of the PostgreSQL database it should be consider to be a breaking change.

## How to test this PR

* Create tests, with and without name server or DS information. Use both `start_domain_test` and `add_batch_job`.
* Verify with `get_test_history` and the filter option that the tests are correctly classified.
* Repeat for all three database engines.
* For PostgreSQL, create the database before the change and use the patch script to update. Test before and after patching.
